src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc

```diff
@@ -0,0 +1,682 @@
+    this->dataPtr->
+        forceMarkerMsg.mutable_material()->mutable_diffuse()->set_b(0);
+    this->dataPtr->
+        forceMarkerMsg.mutable_material()->mutable_diffuse()->set_a(1);
+    this->dataPtr->
+        forceMarkerMsg.mutable_lifetime()->
+          set_sec(this->dataPtr->cameraUpdateRate);
+
+    // Set scales
+    ignition::msgs::Set(this->dataPtr->positionMarkerMsg.mutable_scale(),
+        ignition::math::Vector3d(this->dataPtr->contactRadius,
+        this->dataPtr->contactRadius,
+        this->dataPtr->contactRadius));
+
+    ignition::msgs::Set(this->dataPtr->forceMarkerMsg.mutable_scale(),
+        ignition::math::Vector3d(0.02, 0.02, this->dataPtr->forceLenght));
+}
+
+//////////////////////////////////////////////////
+void OpticalTactilePlugin::PreUpdate(const UpdateInfo &_info,
+                            EntityComponentManager &_ecm)
+{
+    IGN_PROFILE("TouchPluginPrivate::PreUpdate");
+
+    // Nothing left to do if paused
+    if (_info.paused)
+      return;
+
+    if (!this->dataPtr->initialized)
+    {
+        this->dataPtr->Load(_ecm, this->dataPtr->sdfConfig);
+        this->dataPtr->initialized = true;
+    }
+
+    // This is not an "else" because "initialized" can be set
+    // in the if block above
+    if (this->dataPtr->initialized)
+    {
+      // Fetch new collisions from simulation
+      std::vector<Entity> newCollisions;
+      _ecm.Each<components::Collision>(
+          [&](const Entity &_entity, const components::Collision *) -> bool
+          {
+          newCollisions.push_back(_entity);
+          return true;
+          });
+      this->dataPtr->FilterOutCollisions(_ecm, newCollisions);
+
+      // Get the Depth Camera pose
+      ignition::math::Pose3d sensorPose =
+          _ecm.Component<components::Pose>(
+            this->dataPtr->model.Entity())->Data();
+
+      // Depth Camera data is float, so convert Pose3d to Pose3f
+      this->dataPtr->sensorWorldPose = ignition::math::Pose3f(
+          sensorPose.Pos().X(), sensorPose.Pos().Y(), sensorPose.Pos().Z(),
+          sensorPose.Rot().W(), sensorPose.Rot().X(), sensorPose.Rot().Y(),
+          sensorPose.Rot().Z());
+    }
+}
+
+//////////////////////////////////////////////////
+void OpticalTactilePlugin::PostUpdate(
+    const ignition::gazebo::UpdateInfo &_info,
+    const ignition::gazebo::EntityComponentManager &)
+{
+    IGN_PROFILE("TouchPlugin::PostUpdate");
+
+    // Nothing left to do if paused.
+    if (_info.paused)
+      return;
+}
+
+//////////////////////////////////////////////////
+void OpticalTactilePluginPrivate::DepthCameraCallback(
+        const ignition::msgs::PointCloudPacked &_msg)
+{
+  // This check avoids running the callback at t=0 and getting
+  // unexpected markers in the scene
+  if (!this->initialized)
+      return;
+
+  this->cameraMsg = _msg;
+  this->newCameraMsg = true;
+}
+
+//////////////////////////////////////////////////
+void OpticalTactilePluginPrivate::UnpackPointCloudMsg(
```

Comment:
Is it necessary to subscribe to the point cloud message? Processing a point cloud is very time-consuming. Are you using the point cloud just to get the XYZ in the world frame? You should be able to calculate this using the camera matrix. It's a lot faster to do the matrix calculation rather than looping through the point cloud. Matrix speed >> for-loop speed.

Reply:
In `UnpackPointCloudMsg` I am getting the XYZ point in the camera frame corresponding to each (i,j) pixel, out of the message returned by the Depth Camera.

Maybe you are referring to `ComputeNormalForces` method?
Hmm, not sure how to do that, is there any example that implements that?